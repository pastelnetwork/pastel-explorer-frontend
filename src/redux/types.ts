import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { Action } from 'redux';
import { RootState } from '@redux/store';

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export type AppThunkDispatch = ThunkDispatch<RootState, unknown, Action<string>>;
