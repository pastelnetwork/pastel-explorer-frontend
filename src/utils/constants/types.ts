import { LabelKeyObject } from 'react-csv/components/CommonPropTypes';
import { PeriodTypes, TGranularity } from '@utils/helpers/statisticsLib';
import { TStatisticsInfo } from '@utils/types/IStatistics';

export type TThemeColor = {
  name: string;
  backgroundColor: string;
  lineColor?: string;
  splitLineColor: string;
  color: string;
  stack?: string;
  smooth?: boolean;
};

export type TThemeInitOption = {
  theme?: TThemeColor | null;
  data?: number[][];
  dataX?: string[];
  dataY?: number[];
  dataY1?: number[];
  dataY2?: number[];
  chartName: string;
  minY: number;
  maxY: number;
};

export type TLineChartProps = {
  chartName: string;
  period?: PeriodTypes;
  dataX?: string[];
  dataY?: number[];
  dataY1?: number[];
  dataY2?: number[];
  title?: string;
  granularity?: TGranularity;
  info: TStatisticsInfo;
  granularities?: TGranularity[];
  offset: number;
  periods?: PeriodTypes[];
  handleGranularityFilterChange?: (_granularity: TGranularity) => void;
  handleBgColorChange: (_color: string) => void;
  handlePeriodFilterChange?: (_period: PeriodTypes) => void;
};

export type TCsvHeaderType = {
  [key: string]: LabelKeyObject[];
};

export type TScatterChartProps = {
  chartName: string;
  data: number[][];
  dataX: string[];
  title?: string;
  period?: PeriodTypes;
  info: TStatisticsInfo;
  offset: number;
  periods: PeriodTypes[];
  handleBgColorChange: (_color: string) => void;
  handlePeriodFilterChange?: (_period: PeriodTypes) => void;
  handleGranularityFilterChange?: (_granularity: TGranularity) => void;
};
