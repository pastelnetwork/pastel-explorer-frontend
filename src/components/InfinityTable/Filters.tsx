import { memo, FC, useCallback, MouseEvent, useEffect } from 'react';
import makeStyles from '@material-ui/styles/makeStyles';
import Button from '@material-ui/core/Button';
import MenuItem from '@material-ui/core/MenuItem';
import { useDispatch, useSelector } from 'react-redux';
import { setFilterValueAction } from '@redux/actions/filterAction';
import { TAppTheme } from '@theme/index';
import { TFilter } from '@utils/types/IFilter';
import { getFilterState } from '@redux/reducers/filterReducer';

const useStyles = makeStyles((theme: TAppTheme) => {
  return {
    root: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'space-between',
      [theme.breakpoints.down('sm')]: {
        flexDirection: 'column',
      },
    },
    listFilter: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'flex-end',
      border: `1px solid ${theme.palette.background.default}`,
      borderRadius: `${theme.spacing(1)}px`,
    },
    button: {
      paddingRight: 0,
      minWidth: 'auto',
    },
    paper: {
      border: '1px solid',
    },
    rootMenuItem: {
      display: 'block',
      backgroundColor: 'inherit',
      padding: 0,
      minHeight: 'auto',
    },
    rootMenuItemButton: {
      width: '100%',
      textAlign: 'left',
      backgroundColor: 'inherit !important',
      padding: `${theme.spacing(1.5)}px ${theme.spacing(4)}px`,
    },
    rootMenuItemActive: {
      backgroundColor: 'rgba(45, 55, 72, 0.04) !important',
    },
  };
});

interface IProps {
  title: string;
  id?: string;
  filters: TFilter[];
}

const Filters: FC<IProps> = ({ filters, title }) => {
  const dispatch = useDispatch();
  const classes = useStyles();
  const { dateRange } = useSelector(getFilterState);
  const time: string = dateRange || 'all';

  const handleSelectTime = useCallback((event: MouseEvent<HTMLButtonElement>) => {
    const { value } = event.currentTarget;
    dispatch(setFilterValueAction({ dateRange: value }));
  }, []);
  useEffect(() => {
    return () => {
      dispatch(setFilterValueAction({ dateRange: 'all' }));
    };
  }, [dispatch]);
  return (
    <div className={classes.root}>
      <span>{title}</span>
      <div className={classes.listFilter}>
        {filters.map(({ name, value }) => (
          <MenuItem
            key={value}
            classes={{
              root: `${classes.rootMenuItem} ${time === value ? classes.rootMenuItemActive : ''}`,
            }}
          >
            <Button
              classes={{ root: classes.rootMenuItemButton }}
              type="button"
              value={value}
              onClick={handleSelectTime}
            >
              {name}
            </Button>
          </MenuItem>
        ))}
      </div>
    </div>
  );
};

export default memo<IProps>(Filters);
