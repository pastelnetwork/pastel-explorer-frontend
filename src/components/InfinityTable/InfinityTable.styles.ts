import styled from 'styled-components/macro';

import { Card as MuiCard, darken, TableCell, Paper } from '@material-ui/core';
import { spacing } from '@material-ui/system';

const StyledCard = styled(MuiCard)`
  background-color: ${props => props.theme.palette.background.paper} !important;
  padding: 0 ${props => props.theme.spacing(9)}px;
  padding-bottom: 32px !important;
  @media (max-width: 960px) {
    padding: 0 ${props => props.theme.spacing(2)}px;
    padding-bottom: 8px !important;
  }
`;

export const Card = styled(StyledCard)(spacing);

export const TableWrapper = styled.div`
  overflow-y: auto;
  max-width: calc(100vw - ${props => props.theme.spacing(12)}px);
  background-color: ${props => props.theme.palette.background.default} !important;
  border-radius: 16px;

  * {
    outline: 0;
  }
  .ReactVirtualized__Table__headerRow,
  .ReactVirtualized__Grid {
    min-width: 1000px;
    width: 100% !important;
    background-color: ${props => props.theme.palette.background.default} !important;
    &::-webkit-scrollbar-track {
      background: black;
    }

    &::-webkit-scrollbar-track {
      background: grey;
    }
  }

  .ReactVirtualized__Grid__innerScrollContainer {
    max-width: 100% !important;
  }

  .ReactVirtualized__Table__row {
    width: 100% !important;
    background-color: ${props => props.theme.palette.background.default};
    &:nth-of-type(odd): {
      background-color: ${props => darken(props.theme.palette.background.default, 0.5)} !important;
    }
    svg {
      color: ${props => props.theme.palette.text.primary};
    }
  }
  .ReactVirtualized__Table__headerRow,
  .ReactVirtualized__Table__row {
    display: flex;
    transition: all 0.2s ease-in !important;
    &:hover {
      background-color: ${props => darken(props.theme.palette.background.default, 0.07)} !important;
    }
  }
`;

export const TableContainer = styled(Paper)`
  position: relative;
  background: ${props => props.theme.palette.background.default};
`;

export const Cell = styled(TableCell)`
  width: 100%;
  display: flex;
  height: 100%;
  align-items: center;
`;

export const HeaderCell = styled(TableCell)<{ $disabledSort: boolean }>`
  font-weight: 700;
  height: 70px;
  display: flex;
  white-space: nowrap;
  color: ${props => props.theme.palette.text.primary};
  cursor: ${({ $disabledSort }) => !$disabledSort && 'pointer'};
`;

export const Loader = styled.div`
  position: absolute;
  display: flex;
  width: 100%;
  height: 100%;
  background-color: rgba(255, 255, 255, 0.9);
  z-index: 1;
  justify-content: center;
  align-items: center;
`;
