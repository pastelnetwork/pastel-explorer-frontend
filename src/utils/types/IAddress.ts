export interface IAddressData {
  amount: number;
  timestamp: number;
  transactionHash: string;
  direction: string;
}

export interface IAddress {
  address: string;
  incomingSum: number;
  outgoingSum: number;
  data: Array<IAddressData>;
}
