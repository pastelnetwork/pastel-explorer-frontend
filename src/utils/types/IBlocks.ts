export interface IBlockTransaction {
  id: string;
  recipientCount: number;
  totalAmount: number;
}

export interface IBlock {
  confirmations: number;
  difficulty: string;
  height: number;
  id: string;
  merkleRoot: string;
  nextBlockHash: string;
  nonce: string;
  previousBlockHash: string;
  size: number;
  solution: string;
  timestamp: number;
  transactionCount: number;
  transactions: Array<IBlockTransaction>;
}

export interface IRawBlock {
  anchor: string;
  bits: string;
  chainwork: string;
  confirmations: number;
  difficulty: string;
  finalsaplingroot: string;
  hash: string;
  height: number;
  merkleroot: string;
  nonce: string;
  previousblockhash: string;
  size: number;
  solution: string;
  time: number;
  transactions: Array<IBlockTransaction>;
  tx: string[];
}
