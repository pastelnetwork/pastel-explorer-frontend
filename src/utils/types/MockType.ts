import { TFunction } from 'i18next';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type MyMockType = jest.Mock<any, any, any> &
  TFunction<['translation', ...string[]], undefined>;
