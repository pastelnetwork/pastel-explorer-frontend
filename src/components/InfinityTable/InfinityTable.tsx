import * as React from 'react';
import _debounce from 'lodash.debounce';
import {
  AutoSizer,
  Table,
  Column,
  TableCellProps,
  TableHeaderProps,
  ScrollEventData,
  Index,
  OverscanIndicesGetterParams,
} from 'react-virtualized';
import { CSSProperties } from '@material-ui/styles';
import { CardHeader, CircularProgress, darken } from '@material-ui/core';
import { ArrowDropDown, ArrowDropUp } from '@material-ui/icons';
import { Skeleton } from '@material-ui/lab';
import { useGetThemeMode } from '@redux/reducers/appThemeReducer';
import themeVariant from '@theme/variants';
import { TFilter } from '@utils/types/IFilter';
import Filters from './Filters';

import * as Styles from './InfinityTable.styles';

export type HeaderType = { id: number | string; header: string };

export type SortDirectionsType = 'ASC' | 'DESC';

export interface ISortData {
  sortBy: string;
  sortDirection: SortDirectionsType;
  filterBy?: string;
  filterValue?: string;
}

export interface RowsProps {
  [key: string]: number | string | JSX.Element;
}

interface IInfinityTableComponentProps {
  title?: string;
  columns: Array<{
    width: number;
    flexGrow: number;
    label: string;
    dataKey: string;
    disableSort: boolean;
  }>;
  filters?: TFilter[];
  rows: Array<RowsProps>;
  sortBy?: string;
  sortDirection?: SortDirectionsType;
  rowHeight?: number;
  tableHeight?: number;
  loadMoreFrom?: number;
  disableLoading?: boolean;
  renderAllRows?: boolean;
  // eslint-disable-next-line
  onBottomReach?: (value: boolean) => void;
  // eslint-disable-next-line
  onHeaderClick?: (info: ISortData) => void;
}

const noRowsRenderer = () => (
  <Styles.Loader>
    <CircularProgress size={40} />
  </Styles.Loader>
);
const headerRenderer = ({
  label,
  dataKey,
  sortBy,
  sortDirection,
  disableSort,
}: TableHeaderProps & { columnIndex: number }) => {
  const showSort = !disableSort && dataKey === sortBy;
  const renderSortIcon = () =>
    !disableSort && sortDirection === 'DESC' ? <ArrowDropDown /> : <ArrowDropUp />;

  return (
    <Styles.HeaderCell component="div" variant="head" $disabledSort={Boolean(disableSort)}>
      {label}
      {showSort && renderSortIcon()}
    </Styles.HeaderCell>
  );
};
const TableCellRenderer = ({ cellData }: TableCellProps) => (
  <Styles.Cell component="div">{cellData}</Styles.Cell>
);

const getOverscanIndices = ({ cellCount }: OverscanIndicesGetterParams) => ({
  overscanStartIndex: 0,
  overscanStopIndex: cellCount - 1,
});

const InfinityTableComponent: React.FC<IInfinityTableComponentProps> = ({
  title,
  rows,
  columns,
  sortBy,
  filters,
  sortDirection,
  loadMoreFrom = 0,
  onBottomReach,
  onHeaderClick,
  rowHeight = 45,
  tableHeight = 500,
  disableLoading,
  renderAllRows,
}) => {
  const [loading, setLoading] = React.useState(false);
  const isDarkMode = useGetThemeMode();
  const [color, bgOpacity] = React.useMemo(
    () =>
      isDarkMode ? [themeVariant.custom.blue.blackRock, 0.5] : [themeVariant.custom.white, 0.05],
    [isDarkMode],
  );
  const handleReachBottom = _debounce(
    ({ clientHeight, scrollHeight, scrollTop }: ScrollEventData) => {
      if (!onBottomReach || rows.length < loadMoreFrom) return null;

      const bottomReached = clientHeight + scrollTop >= scrollHeight;
      !loading && bottomReached && setLoading(true);

      return onBottomReach(bottomReached);
    },
    100,
  );

  const handleSort = (info: ISortData) => {
    !loading && !disableLoading && setLoading(true);

    return onHeaderClick && onHeaderClick(info);
  };

  const handleRowStyle = ({ index }: Index): CSSProperties => {
    if (index % 2 === 0) {
      return {
        backgroundColor: darken(color, bgOpacity),
      };
    }

    return {
      backgroundColor: color,
    };
  };

  React.useEffect(() => {
    loading && setLoading(false);
  }, [rows]);

  return (
    <Styles.Card mb={3}>
      {title && (
        <CardHeader title={!filters ? title : <Filters filters={filters} title={title} />} />
      )}
      <Styles.TableContainer>
        {loading && (
          <Styles.Loader>
            <CircularProgress size={40} />
          </Styles.Loader>
        )}
        {rows.length ? (
          <Styles.TableWrapper>
            <AutoSizer disableHeight>
              {({ width }) => (
                <div style={{ width }}>
                  <Table
                    headerHeight={rowHeight}
                    height={tableHeight}
                    noRowsRenderer={noRowsRenderer}
                    rowHeight={rowHeight}
                    rowGetter={({ index }: { index: number }) => rows[index]}
                    rowCount={rows.length}
                    rowStyle={(info: Index) => handleRowStyle(info)}
                    width={1920}
                    sortBy={sortBy}
                    sort={handleSort}
                    sortDirection={sortDirection}
                    style={{ backgroundColor: 'inherit' }}
                    onScroll={handleReachBottom}
                    overscanIndicesGetter={renderAllRows ? getOverscanIndices : undefined}
                  >
                    {columns.map(({ dataKey, ...other }, index) => (
                      <Column
                        key={dataKey}
                        headerRenderer={headerProps =>
                          headerRenderer({
                            ...headerProps,
                            columnIndex: index,
                          })
                        }
                        cellRenderer={TableCellRenderer}
                        dataKey={dataKey}
                        {...other}
                      />
                    ))}
                  </Table>
                </div>
              )}
            </AutoSizer>
          </Styles.TableWrapper>
        ) : (
          <Skeleton animation="wave" variant="rect" height={tableHeight} />
        )}
      </Styles.TableContainer>
    </Styles.Card>
  );
};

export default InfinityTableComponent;
