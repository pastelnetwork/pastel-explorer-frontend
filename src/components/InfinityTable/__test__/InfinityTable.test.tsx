import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import 'jest-styled-components';

import { MyMockType } from '@utils/types/MockType';
import i18next from '../../../utils/helpers/i18n';
import InfinityTable from '../InfinityTable';

const mockStore = configureMockStore();
const store = mockStore({});

jest.mock('i18next-http-backend');
jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            // noop
          }),
      },
    };
  },
  initReactI18next: {
    type: '3rdParty',
    init: () => {
      // noop
    },
  },
}));
i18next.t = jest.fn().mockImplementation((...arg) => {
  return arg[0];
}) as MyMockType;

const columns = [
  {
    width: 100,
    flexGrow: 1,
    label: 'ID',
    dataKey: 'id',
    disableSort: false,
  },
  {
    width: 200,
    flexGrow: 1,
    label: 'Name',
    dataKey: 'name',
    disableSort: false,
  },
];
const rows = [
  {
    id: 'd6f9484a-c3a3-11ed-afa1-0242ac120002',
    name: 'David',
  },
  {
    id: 'f05836a2-c3a3-11ed-afa1-0242ac120002',
    name: 'Henry',
  },
];

describe('components/InfinityTable', () => {
  const wrapper = shallow(
    <Provider store={store}>
      <InfinityTable columns={columns} rows={rows} />
    </Provider>,
  );

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
