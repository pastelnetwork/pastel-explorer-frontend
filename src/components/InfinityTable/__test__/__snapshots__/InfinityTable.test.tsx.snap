// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/InfinityTable renders correctly 1`] = `ShallowWrapper {}`;
